### ISSUE TITLE
Undefined variable: event aka PHP sessions mismatch 

### ISSUE THREAD

**Issue by drzraf** [trilbymedia/grav-plugin-login-oauth2#11]
While trying to connect to `/admin/`:

```
Whoops\Exception\ErrorException thrown with message "Undefined variable: event"

Stacktrace:
#35 Whoops\Exception\ErrorException in /grav/mx/plugins/login-oauth2/login-oauth2.php:224
#34 Grav\Common\Debugger:deprecatedErrorHandler in /grav/mx/plugins/login-oauth2/login-oauth2.php:224
#33 Grav\Plugin\LoginOauth2Plugin:loginCallback in /grav/vendor/symfony/event-dispatcher/EventDispatcher.php:212
#32 Symfony\Component\EventDispatcher\EventDispatcher:doDispatch in /grav/vendor/symfony/event-dispatcher/EventDispatcher.php:44
#31 Symfony\Component\EventDispatcher\EventDispatcher:dispatch in /grav/vendor/rockettheme/toolbox/Event/src/EventDispatcher.php:23
#30 RocketTheme\Toolbox\Event\EventDispatcher:dispatch in /grav/system/src/Grav/Common/Grav.php:365
#29 Grav\Common\Grav:fireEvent in /grav/system/src/Grav/Common/Processors/TasksProcessor.php:52
#28 Grav\Common\Processors\TasksProcessor:process in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:45
#27 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:57
#26 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Common/Processors/RequestProcessor.php:49
#25 Grav\Common\Processors\RequestProcessor:process in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:45
#24 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:57
#23 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Common/Processors/ThemesProcessor.php:27
#22 Grav\Common\Processors\ThemesProcessor:process in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:45
#21 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:57
#20 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Common/Processors/PluginsProcessor.php:30
#19 Grav\Common\Processors\PluginsProcessor:process in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:45
#18 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:57
#17 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Common/Processors/InitializeProcessor.php:78
#16 Grav\Common\Processors\InitializeProcessor:process in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:45
#15 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:57
#14 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Common/Processors/DebuggerProcessor.php:27
#13 Grav\Common\Processors\DebuggerProcessor:process in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:45
#12 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:57
#11 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Common/Processors/ErrorsProcessor.php:27
#10 Grav\Common\Processors\ErrorsProcessor:process in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:45
#9 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:57
#8 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Common/Processors/LoggerProcessor.php:48
#7 Grav\Common\Processors\LoggerProcessor:process in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:45
#6 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:57
#5 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Common/Processors/ConfigurationProcessor.php:28
#4 Grav\Common\Processors\ConfigurationProcessor:process in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:45
#3 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Framework/RequestHandler/Traits/RequestHandlerTrait.php:57
#2 Grav\Framework\RequestHandler\RequestHandler:handle in /grav/system/src/Grav/Common/Grav.php:249
#1 Grav\Common\Grav:process in /grav/index.php:53
#0 require in /grav/system/router.php:27
```

### Context

- `$event` is undefined.
- `$state` is undefined.
- `$provider_name` and `$this->isValidProvider($provider_name)` are respectively `null` and `false`.
- `$session->oauth2_provider` does not exist
- `$session` is
```
object(Grav\Common\Session)[134]
  protected 'autoStart' => boolean false
  protected 'options' => 
    array (size=11)
      'name' => string 'grav-site-023cd39' (length=17)
      'cookie_lifetime' => string '1800' (length=4)
      'cookie_path' => string '/' (length=1)
      'cookie_domain' => string '' (length=0)
      'cookie_secure' => string '0' (length=1)
      'cookie_httponly' => string '1' (length=1)
      'cache_limiter' => string 'nocache' (length=7)
      'use_trans_sid' => string '0' (length=1)
      'use_cookies' => string '1' (length=1)
      'lazy_write' => string '1' (length=1)
      'use_strict_mode' => string '1' (length=1)
  protected 'started' => boolean true
```

Not yet sure, but I think it's related to a point I talked about in #7 about cookie domain being somehow inconsistent. Quoting:
> session is not loaded from the session id stored by the grav-site-XXX-admin cookie, but the grav-site-XXX cookie which does not contain the necessary information.

**Comment by rhukster** [comment]
I have this fixed locally... but trying to find a better solution for handling site vs. admin configurations for diff oauth providers.

**Comment by drzraf** [comment]
I tried without `session split` + the missing `$session->init();` and got better results.
But another problem is that `addEnabledProviders` is not instanciated using `$admin = true` although it should.

**Comment by drzraf** [comment]
I very pleased I finally could get it working with `session.split = false` in combination with this patch:
```diff
diff --git a/classes/OAuth2.php b/classes/OAuth2.php
index 12bacab..6d52f19 100644
--- a/classes/OAuth2.php
+++ b/classes/OAuth2.php
@@ -27,7 +27,7 @@ public function isAdmin()
 
     public function addEnabledProviders()
     {
-        if ($this->admin) {
+        if (true) {
             $providers = isset($this->config['admin']['providers']) ? (array)$this->config['admin']['providers'] : [];
         } else {
             $providers = isset($this->config['providers']) ? (array)$this->config['providers'] : [];
diff --git a/classes/Providers/BaseProvider.php b/classes/Providers/BaseProvider.php
index ca0fde7..075cdf6 100644
--- a/classes/Providers/BaseProvider.php
+++ b/classes/Providers/BaseProvider.php
@@ -28,7 +28,7 @@
      */
     public function __construct()
     {
-        $admin = Grav::instance()['oauth2']->isAdmin();
+        $admin = true;
         $this->config = new Data(Grav::instance()['config']->get('plugins.login-oauth2' . ($admin ? '.admin' : '')));
         $this->state = 'LOGIN_OAUTH2_' . Utils::generateRandomString(15);
     }
diff --git a/composer.json b/composer.json
index 1131443..296924a 100644
--- a/composer.json
+++ b/composer.json
@@ -4,7 +4,7 @@
         "league/oauth2-github": "^2.0",
         "league/oauth2-instagram": "^2.0",
         "league/oauth2-facebook": "^2.0",
-        "league/oauth2-google": "^2.0",
+        "league/oauth2-google": "3.0.1",
         "league/oauth2-linkedin": "^3.0"
     },
     "autoload": {
diff --git a/login-oauth2.php b/login-oauth2.php
index 13019cc..00736f2 100644
--- a/login-oauth2.php
+++ b/login-oauth2.php
@@ -175,6 +175,7 @@ public function loginCallback()
 
         /** @var Session $session */
         $session = $this->grav['session'];
+        $session->init();
         $provider_name = $session->oauth2_provider;
 
         /** @var Language $t */
```

It's not mergeable by itself but highlights the two main problems:
1. session initialization
2. In the plugin configuration, `admin configuration domain` must be being correctly set in at least two additional places. For this, `OAuth2` class may need a `setAdmin()` method which is called according to whether `loginCallback` is used for **admin** authentication.

**Comment by rhukster** [comment]
As I said, I was working on on this, and i'm getting ready to release a version 2.0 that has been completely refactored to support the split sessions that are standard in Grav.  I suggest trying with the `develop` branch.

**Comment by rhukster** [comment]
Is this based on the current `develop` branch or released version??

**Comment by rhukster** [comment]
I can't replicate this with latest `develop` branch.. You need to have both Site and Admin settings the same though.

![Image 2019-04-26 at 2 08 51 PM](https://user-images.githubusercontent.com/1084697/56833822-df8e3780-682c-11e9-8197-bdbaee661cf8.png)

**Comment by drzraf** [comment]
Indeed, I didn't set the `site` part because I'm just interested in admin (and would rather avoid duplicating credentials)

**Comment by rhukster** [comment]
I've tested with just setting the admin stuff (site side disabled) and it works fine split or not split.  You do need to have the right admin callback set though.

**Comment by drzraf** [comment]
I got it working! Having a **different /admin/ callback** was a key here. By defining it to a different URL, `admin` is correctly set, session and providers initialized, ... and indeed, in both `split` and `non-split` session modes (and since `admin` is correctly set, duplicating into `site settings` isn't even needed).
Thank you!

Out of curiosity, don't you think there is possibility to have only one callback URL for both systems or is there a limitation I overlooked about determining the *intended use* of the callback in order to select the *session* ?
